### PR TITLE
add comma operator to the deprecated list

### PR DESCRIPTION
**Pull request by wilzbach**
As we are finally getting rid of the comma operator, how about adding it to the "Future" list. Once the [PR](https://github.com/dlang/dmd/pull/5737) is merged, more infos can be added?
Ping @Geod24 @yebblies

### CODE REVIEW

File: deprecate.dd [hunk 1]
```diff
@@ -55,6 +56,32 @@ $(SPEC_S Deprecated Features,
...
+
+        ---
+        for (; !a.empty && !b.empty; a.popFront, b.popFront)
+        ---
+    )
+$(H4 Corrective Action)
+    $(P Usage of the comma operator outside of for-loops should be forbidden.
+    )
+$(H4 Rationale)
+    $(P The comma operator leads to a lot of false friends,
```
Comment (Hackerpilot) [inline deprecate.dd]: Maybe rephrase as "unintended behavior"

File: deprecate.dd [hunk 2]
```diff
@@ -55,6 +56,32 @@ $(SPEC_S Deprecated Features,
     )
 
 
+$(H3 $(DEPNAME Comma operator))
+    $(P The comma operator (`,`) allows to execute multiple expressions and
```
Comment (Hackerpilot) [inline deprecate.dd]: "allows executing multiple"

File: deprecate.dd [hunk 3]
```diff
@@ -55,6 +56,32 @@ $(SPEC_S Deprecated Features,
...
+        ---
+    )
+$(H4 Corrective Action)
+    $(P Usage of the comma operator outside of for-loops should be forbidden.
+    )
+$(H4 Rationale)
+    $(P The comma operator leads to a lot of false friends,
+        (e.g. `synchronized (lockA, lockB)` isn't currently implemented, but still
+        compiles "thanks" to the comma operator).
+        Moreover it has a tiny (mostly accidental) usage
```
Comment (Hackerpilot) [inline deprecate.dd]: Maybe "Moreover it is not commonly used and it blocks the ability to implement tuples"

**Comment by wilzbach** [comment]
Thanks @Hackerpilot - addressed and rebased.

File: deprecate.dd [hunk 4]
```diff
@@ -55,6 +56,31 @@ $(SPEC_S Deprecated Features,
     )
 
 
+$(H3 $(DEPNAME Comma operator))
+    $(P The comma operator (`,`) allows executing multiple expressions and
+        discards them except for the last which is returned.
```
Comment (yebblies) [inline deprecate.dd]: The expressions aren't discarded, only their results.

File: deprecate.dd [hunk 5]
```diff
@@ -55,6 +56,31 @@ $(SPEC_S Deprecated Features,
...
+        ---
+        int a = 1;
+        int b = 2;
+        bool ret = a == 2, b == 2; // true
+        ---
+
+        It's also common to use the comma operator in for-loop increment
+        statements to allow multiple statements.
```
Comment (yebblies) [inline deprecate.dd]: multiple expressions, not statements.

File: deprecate.dd [hunk 6]
```diff
@@ -55,6 +56,31 @@ $(SPEC_S Deprecated Features,
...
+
+        ---
+        for (; !a.empty && !b.empty; a.popFront, b.popFront)
+        ---
+    )
+$(H4 Corrective Action)
+    $(P Usage of the comma operator outside of for-loops should be forbidden.
```
Comment (yebblies) [inline deprecate.dd]: This is not the path we're going with.  Please take a look at the current PRs.

File: deprecate.dd [hunk 7]
```diff
@@ -55,6 +56,31 @@ $(SPEC_S Deprecated Features,
...
+    )
+$(H4 Rationale)
+    $(P The comma operator leads to unintended behavior,
+        (e.g. `synchronized (lockA, lockB)` isn't currently implemented, but still
+        compiles "thanks" to the comma operator).
```
Comment (yebblies) [inline deprecate.dd]: This is not the best example, since it's actually a parser bug.  There are plenty of more common examples.

Comment (Hackerpilot) [inline deprecate.dd]: No it's not. The grammar says:
```
SynchronizedStatement:
    synchronized ( Expression ) ScopeStatement
Expression:
    CommaExpression
```
It's perfectly valid for a D parser to accept this code.

Comment (yebblies) [inline deprecate.dd]: I feel like we've had this discussion before and you probably know what I mean.  Spec bug, parser bug, whatever.  The point is that even with comma expressions still allowed in the language we could avoid accepting this bug by changing the parser.

Comment (wilzbach) [inline deprecate.dd]: > There are plenty of more common examples.

I went through the github issue and found those:

```
writeln( 6, mixin("7,8"), 9 ); // 6, 8, 9
vec v = (0, 0, 3); // 3

int b = 2;
if (a == 1, b == 2) {
  // will always be reached
}
```

do you have better ones in mind? Just pick an example ;-)

File: deprecate.dd [hunk 8]
```diff
@@ -55,6 +56,31 @@ $(SPEC_S Deprecated Features,
...
+
+        It's also common to use the comma operator in for-loop increment
+        statements to allow multiple expressions.
+
+        ---
+        for (; !a.empty && !b.empty; a.popFront, b.popFront)
+        ---
+    )
+$(H4 Corrective Action)
+    $(P Using the result of the comma operator should be forbidden.
```
Comment (wilzbach) [inline deprecate.dd]: @yebblies did I formulate the action correctly?
Isn't it still planned to deprecate comma operator on the long run and replace it with tuples?

**Comment by mathias-lang-sociomantic** [comment]
Regarding the 'needs approval', I believe this change already have been approved: https://forum.dlang.org/post/nh1luh$20ds$1@digitalmars.com

File: deprecate.dd [hunk 9]
```diff
@@ -10,6 +10,7 @@ $(SPEC_S Deprecated Features,
 
     $(TABLE2 Deprecated Features,
         $(THEAD Feature,                                                          Spec,  Dep,    Error,  Gone)
+        $(TROW $(DEPLINK Comma operator),                                         future, &nbsp;, &nbsp;, &nbsp;)
```
Comment (mathias-lang-sociomantic) [inline deprecate.dd]: "Using the result of a comma expression"
We can extend it later, but so far that's what we are discussing.

**Comment by wilzbach** [comment]
@mathias-lang-sociomantic I updated the title and added a list of unintended behaviors with the comma operator - I hope everyone can find his favorite bug in there ;-)

File: deprecate.dd [hunk 10]
```diff
@@ -55,6 +56,48 @@ $(SPEC_S Deprecated Features,
...
+        ---
+    )
+$(H4 Corrective Action)
+    $(P Using the result of the comma operator should be forbidden.
+    )
+$(H4 Rationale)
+    $(P The comma operator leads to unintended behavior (see below for a selection)
+        Moreover it is not commonly used and it blocks the ability to implement tuples.
+
+        A selection of problems through the accidental use of the comma operator:
```
Comment (wilzbach) [inline deprecate.dd]: is this line redundant?

**Comment by wilzbach** [comment]
So if this is already approved, we can move forward?
Any other comments?

File: deprecate.dd [hunk 11]
```diff
@@ -55,6 +56,48 @@ $(SPEC_S Deprecated Features,
...
+
+        It's also common to use the comma operator in for-loop increment
+        statements to allow multiple expressions.
+
+        ---
+        for (; !a.empty && !b.empty; a.popFront, b.popFront)
+        ---
+    )
+$(H4 Corrective Action)
+    $(P Using the result of the comma operator should be forbidden.
```
Comment (yebblies) [inline deprecate.dd]: This should be a description of how to fix code - ie replace with a lambda or statement.  Check the other items on this page for other examples of what the Corrective Action section is for.

**Comment by yebblies** [comment]
Apart from that looks ok.

**Comment by wilzbach** [comment]
> Apart from that looks ok.

rebased.

**Comment by wilzbach** [comment]
@yebblies @mathias-lang-sociomantic more nitpicks or are we ready to go?

**Comment by yebblies** [comment]
Yup LGTM.  I don't have merge rights here or I'd merge it.
